modules/http/http-grammar.ym

```diff
@@ -129,6 +158,7 @@ http_option
     | http_tls_option
     | KW_TLS '(' http_tls_options ')'
     | { last_template_options = http_dd_get_template_options(last_driver); } template_option
+    | KW_ON_HTTP_ERROR on_error_block
```

Comment:
I would like to have the grammar improved somewhat. The response code from the server does not necessarily indicate an error, in http terms the code is a "status" code. Maybe even using the "arrow" syntax that we introduced for name-value pairs would be useful here.

Something like this:

```
http(.... 
         handle-response(404 => drop,
                                     418 => disconnect)
```

Reply:
I would like to leave some way to further expand the match logic. The match logic might depend on the headers of the exact content of the http response. In my original branch, I had an additional error-msg(string) that would allow users to further branch per status code, based on the body of the response. I did not send that part in the end because it caused significant code complexity, and was not necessary for the resolution of #2996 . But later we might want to add such thing incrementally. That's why I used the current syntax.

Still, currently there is an unused user_data passed into the action callbacks. That could be later used to transfer the header and body content. So we can say the logic that further dispatches must be called inside the callbacks. With that we can save the `=>` syntax.

Which one is preferable in your opinion?

Reply:
I don't think we should branch on anything but the status code. REST style APIs uses this code to convey information about what has happened to the request.

Headers and response contents could be useful for a more active logic, but I don't think those should be available via configuration. I can imagine plugins on top of http() that processes the response in depth (e.g. consuming a per-message status/acknowledgement from the server returned via  JSON array in the response), but I think should we want to support that we would need to implement this via a higher-order plugin: one that can potentially be written in Python or another language. The "event" mechanism that we discussed for file-rotation could facilitate something like this in the future.

Reply:
I understand and agree. I updated the PR with the `=>` syntax.